dask_kubernetes/operator/daskcluster.py

```diff
@@ -0,0 +1,401 @@
+        },
+        "spec": {
+            "containers": [
+                {
+                    "name": "scheduler",
+                    "image": image,
+                    "args": [
+                        "dask-worker",
+                        f"tcp://{scheduler_name}.{namespace}:8786",
+                        f"--name={worker_name}",
+                    ],
+                }
+            ]
+        },
+    }
+
+
+def build_worker_group_spec(name, image, replicas, resources, env):
+    return {
+        "apiVersion": "kubernetes.dask.org/v1",
+        "kind": "DaskWorkerGroup",
+        "metadata": {"name": f"{name}-worker-group"},
+        "spec": {
+            "image": image,
+            "replicas": replicas,
+            "resources": resources,
+            "env": env,
+            "minimum": replicas,
+            "maximum": replicas,
+        },
+    }
+
+
+def build_cluster_spec(name, image, replicas, resources, env):
+    return {
+        "apiVersion": "kubernetes.dask.org/v1",
+        "kind": "DaskCluster",
+        "metadata": {"name": f"{name}-cluster"},
+        "spec": {
+            "image": image,
+            "scheduler": {"serviceType": "ClusterIP"},
+            "replicas": replicas,
+            "resources": resources,
+            "env": env,
+        },
+    }
+
+
+def build_cluster_spec(name, image, replicas, resources, env):
+    return {
+        "apiVersion": "kubernetes.dask.org/v1",
+        "kind": "DaskCluster",
+        "metadata": {"name": f"{name}-cluster"},
+        "spec": {
+            "image": image,
+            "scheduler": {"serviceType": "ClusterIP"},
+            "replicas": replicas,
+            "resources": resources,
+            "env": env,
+        },
+    }
+
+
+async def wait_for_scheduler(cluster_name, namespace):
+    async with kubernetes.client.api_client.ApiClient() as api_client:
+        api = kubernetes.client.CoreV1Api(api_client)
+        watch = kubernetes.watch.Watch()
+        async for event in watch.stream(
+            func=api.list_namespaced_pod,
+            namespace=namespace,
+            label_selector=f"dask.org/cluster-name={cluster_name},dask.org/component=scheduler",
+            timeout_seconds=60,
+        ):
+            if event["object"].status.phase == "Running":
+                watch.stop()
+            await asyncio.sleep(0.1)
+
+
+async def get_scheduler_address(service_name, namespace):
+    async with kubernetes.client.api_client.ApiClient() as api_client:
+        api = kubernetes.client.CoreV1Api(api_client)
+    service_name = "foo-cluster"
+    service = await api.read_namespaced_service(service_name, namespace)
+    port_forward_cluster_ip = None
+    address = await get_external_address_for_scheduler_service(
+        api, service, port_forward_cluster_ip=port_forward_cluster_ip
+    )
+    return address
+
+
+@kopf.on.create("daskcluster")
+async def daskcluster_create(spec, name, namespace, logger, **kwargs):
+    global SCHEDULER_NAME
```

Comment:
We need to switch this out for some other way of handling these references. Perhaps a singleton class?